File: React_Implementation/grocery-app/frontend/src/services/userService.js
import api, { firstSuccess } from './api';

export const userService = {
  createUser: async (user) => (await firstSuccess([
    () => api.post('/api/users', user),
    () => api.post('/users', user),
  ])).data,
  getAllUsers: async () => (await firstSuccess([
    () => api.get('/api/users'),
    () => api.get('/users'),
  ])).data,
  getUserById: async (id) => (await firstSuccess([
    () => api.get(`/api/users/${id}`),
    () => api.get(`/users/${id}`),
  ])).data,
  updateUser: async (id, user) => (await firstSuccess([
    () => api.put(`/api/users/${id}`, user),
    () => api.put(`/users/${id}`, user),
  ])).data,
  deleteUser: async (id) => (await firstSuccess([
    () => api.delete(`/api/users/${id}`),
    () => api.delete(`/users/${id}`),
  ])).data,
  searchByEmail: async (email) => {
    const encoded = encodeURIComponent(email);
    return (await firstSuccess([
      () => api.get(`/api/users/search/email/${encoded}`),
      () => api.get(`/api/users/search?email=${encoded}`),
      () => api.get(`/users/search/email/${encoded}`),
    ])).data;
  },
  searchByName: async (name) => {
    const encoded = encodeURIComponent(name);
    return (await firstSuccess([
      () => api.get(`/api/users/search/name/${encoded}`),
      () => api.get(`/api/users/search?name=${encoded}`),
      () => api.get(`/users/search/name/${encoded}`),
    ])).data;
  },
};

File: Docker-Implementation/grocery-app/backend/src/main/java/com/example/groceryapp/service/UserService.java
package com.example.groceryapp.service;

import com.example.groceryapp.entity.User;
import com.example.groceryapp.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    // CREATE
    public User createUser(User user) {
        return userRepository.save(user);
    }

    // READ ALL
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    // READ BY ID
    public Optional<User> getUserById(Integer id) {
        return userRepository.findById(id);
    }

    // UPDATE
    public User updateUser(Integer id, User userDetails) {
        User user = userRepository.findById(id).orElse(null);

        if (user != null) {
            user.setName(userDetails.getName());
            user.setEmail(userDetails.getEmail());
            user.setPassword(userDetails.getPassword());
            user.setPhoneNo(userDetails.getPhoneNo());
            user.setAddress(userDetails.getAddress());
            user.setRole(userDetails.getRole());
            return userRepository.save(user);
        }
        return null;
    }

    // DELETE
    public void deleteUser(Integer id) {
        userRepository.deleteById(id);
    }

    // SEARCH BY NAME
    public List<User> searchByName(String name) {
        return userRepository.findByNameContaining(name);
    }

    // SEARCH BY EMAIL
    public Optional<User> searchByEmail(String email) {
        return userRepository.findByEmail(email);
    }
}
